crates/goose/src/agents/prompt_manager.rs

```diff
@@ -0,0 +1,97 @@
+// agent/prompt_builder.rs
+use chrono::Utc;
+use serde_json::Value;
+use std::collections::HashMap;
+
+use crate::agents::extension::ExtensionInfo;
+use crate::{config::Config, prompt_template};
+
+pub struct PromptManager {
+    system_prompt_override: Option<String>,
+    system_prompt_extras: Vec<String>,
+}
+
+impl Default for PromptManager {
+    fn default() -> Self {
+        PromptManager::new()
+    }
+}
+
+impl PromptManager {
+    pub fn new() -> Self {
+        PromptManager {
+            system_prompt_override: None,
+            system_prompt_extras: Vec::new(),
+        }
+    }
+
+    /// Add an additional instruction to the system prompt
+    pub fn add_system_prompt_extra(&mut self, instruction: String) {
+        self.system_prompt_extras.push(instruction);
+    }
+
+    /// Override the system prompt with custom text
+    pub fn set_system_prompt_override(&mut self, template: String) {
+        self.system_prompt_override = Some(template);
+    }
+
+    /// Build the final system prompt
+    ///
+    /// * `extensions_info` – extension information for each extension/MCP
+    /// * `frontend_instructions` – instructions for the "frontend" tool
+    pub fn build_system_prompt(
+        &self,
+        extensions_info: Vec<ExtensionInfo>,
+        frontend_instructions: Option<String>,
+    ) -> String {
+        let mut context: HashMap<&str, Value> = HashMap::new();
+        let mut extensions_info = extensions_info.clone();
+
+        // Add frontend tools as a special extension if any exist
+        let frontend_instructions = frontend_instructions.unwrap_or(
+            "The following tools are provided directly by the frontend and will be executed by the frontend when called.".to_string()
+        );
+        extensions_info.push(ExtensionInfo::new(
+            "frontend",
+            &frontend_instructions,
+            false,
+        ));
```

Comment:
shouldn't this be an `if let Some(frontend_instructions) = frontend_instructions {...}`? 
we would only want to add `frontend` extension info if this is `Some` and not default to adding it in the `None` case

Reply:
good catch!